File: mergeKSortedList.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 * int val;
 * ListNode next;
 * ListNode() {}
 * ListNode(int val) { this.val = val; }
 * ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
// Time Complexity : O(nklogk), k is the number of list and n is the size of the
// list
// Space Complexity : O(k)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

// Your code here along with comments explaining your approach in three
// sentences only
/*
 * Maintaina a minHeap to store all lists in the queue.
 * We get the minimum value and append it to the result and if the next element
 * for the list is not emepty add it to the queue,
 * Run this till the queue is not empty
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        ListNode result = new ListNode(-1);
        ListNode current = result;

        PriorityQueue<ListNode> pq = new PriorityQueue<>((a, b) -> a.val - b.val);

        for (ListNode list : lists) {
            if (list != null) {
                pq.add(list);
            }
        }

        while (!pq.isEmpty()) {
            ListNode minimum = pq.poll();
            current.next = minimum;
            current = current.next;

            if (minimum.next != null) {
                pq.add(minimum.next);
            }
        }

        return result.next;
    }
}

File: MergeKSortedList.cpp
//Time Complexity O(nlogn)
// Space Complexity O(n)
//Problem successfully executed on leetcode
//No problems faced while coading this.


#include<iostream>
#include<vector>
#include<queue>
#include<map>
#include<stack>
using namespace std;





  struct ListNode {
      int val;
      ListNode *next;
      ListNode() : val(0), next(nullptr) {}
     ListNode(int x) : val(x), next(nullptr) {}
     ListNode(int x, ListNode *next) : val(x), next(next) {}
  };
 
struct compare{
  bool operator() (ListNode* a, ListNode* b){
      return a->val > b->val;
  }
};

class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        priority_queue<ListNode*, vector<ListNode*>, compare> pq;
        
        for(auto &list:lists)
        {
            if(list!=NULL)
            {
                pq.push(list);
            }
        }
        
        ListNode* dummy= new ListNode(-1);
        ListNode* current=dummy;
        
        
        while(!pq.empty())
        {
            ListNode* minNode= pq.top();
            pq.pop();
            if(minNode->next!=NULL)
            {
                pq.push(minNode->next);
            }
            current->next=minNode;
            current=minNode;
        }
        return dummy->next;
    }
};